Auto resume trigger breakpoints once hit

SougandhS opened the pull request:
This PR introduces an option to resume the virtual machine execution once a trigger breakpoint is hit. This allows trigger breakpoints to act as barriers for other breakpoints, enabling less priority breakpoints to remain active without requiring them to be disabled.

Screens - 
<img width="520" alt="image" src="https://github.com/user-attachments/assets/172dc86b-9144-4d7a-a2b7-00a902ae09c0" />
<br>Once trigger is enabled then users will be able to choose 'Resume' , Selecting 'Resume' will also include 'Ignored' tag to breakpoint label
<img width="541" alt="image" src="https://github.com/user-attachments/assets/391b6a81-7c7b-477b-b90b-c67c3c7e3a33" />
<br> On debugging trigger wont suspend at breakpoint line, it will only suspend on next breakpoint

_________________________________________
<br><br> Additionally a new option has been added to the Java editor to toggle Triggerpoints, allowing users to
instantly create a Resume-Triggerpoint. <br>
<img width="483" alt="Screenshot 2025-02-24 at 1 33 24 PM" src="https://github.com/user-attachments/assets/9dfd35ab-b11a-4c93-b1dc-960df36d2b4b" />
<br>

Fixes https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/614

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/eclipse-jdt/.github/blob/main/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See https://github.com/eclipse-jdt/.github/security/policy
-->

## What it does
<!-- Include relevant issues and describe how they are addressed. -->

## How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

SougandhS commented:
Hi @jukzi , could you pls check this when you have time..

jukzi commented:
> Hi @jukzi , could you pls check this when you have time..

Code looks fine. I am testing it, thanks for working on that!! And congrats to your Election.

jukzi commented:
It works almost like i wished it would.
I see a single unexpected behaviour:
Lets have a Resume point with a Condition. When i edit the Condition the Breakpoint disables its "Resume" behavior back to single thread (may be an existing bug):
![image](https://github.com/user-attachments/assets/21d20da2-17d1-417d-a99c-c07a35f41ced)
![image](https://github.com/user-attachments/assets/5803108d-939c-48fd-be54-6e2a767440a9)

Also i would like to see a "Toggle Trigger Point" in the context menu, which would "Resume" *by default* (debatable) :
![image](https://github.com/user-attachments/assets/5f6d1e55-c9cb-4bb4-b5f8-51d8cdf49dab)

jukzi commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/debug/core/IJavaBreakpoint.java:
```diff
@@ -47,6 +47,12 @@ public interface IJavaBreakpoint extends IBreakpoint, ITriggerPoint {
+	/**
+	 * Suspend policy constant indicating a breakpoint will not suspend the target VM when hit.
+	 *
+	 * @since 3.5
```
expecting `@since 3.22`
```
00:01:32.006  [ERROR] Failed to execute goal org.eclipse.tycho:tycho-apitools-plugin:4.0.11:verify (verify) on project org.eclipse.jdt.debug: There are API errors:
00:01:32.006  [ERROR] model/org/eclipse/jdt/debug/core/IJavaBreakpoint.java:54 Invalid @since 3.5 tag on RESUME_THREAD; expecting @since 3.22
```

note that this  change has to wait for M1 anyway and then you may need to bump one more time

jukzi reviewed (dismissed):
fix API error

SougandhS commented:
> When i edit the Condition the Breakpoint disables its "Resume" behavior back to single thread 

Will check this

> Also i would like to see a "Toggle Trigger Point" in the context menu, which would "Resume" by default (debatable) :

Sure 👍

SougandhS commented:
Added additional handling for conditional-resume-triggerpoints

Once resume is enabled, resume based labels will be shown for condition options

<img width="530" alt="Screenshot 2025-02-24 at 1 27 24 PM" src="https://github.com/user-attachments/assets/fc7d0ac3-541a-404a-8119-b4c2df348a83" />

Added toggling Trigger points from context menu (by default it will be Resume-Triggers) <br>
<img width="483" alt="image" src="https://github.com/user-attachments/assets/b1fc920a-42d3-4b28-90ca-27048fbaaff4" />

jukzi commented:
Take this example:
```
package p1;

public class TestResume {
public static void main(String[] args) throws InterruptedException {
	System.out.println("main");
	Thread thread = new Thread(TestResume::runAsync);
	thread.start(); // add Triggerpoint with condition=false
	thread.join();
}
public static void runAsync() {
	System.out.println("async"); // add breakpint
}
}
```
and add the breakpoint as in the comments:
![image](https://github.com/user-attachments/assets/0fcf08c4-dec0-4dd1-b91c-c09f8ca048ba)

I found two problems:
Error 1: The Triggerpoint stops the program (should not happen, its a Triggerpoint, not a Breakpoint)
Error 2: The program never ends - even if i press Resume(F8). Thats fatal

SougandhS commented:
>Take this example:

thanks for this example
>Error 1: The Triggerpoint stops the program (should not happen, its a Triggerpoint, not a Breakpoint)
>Error 2: The program never ends - even if i press Resume(F8). Thats fatal

fixed both now 👍

iloveeclipse commented:
I'm still in the middle of code review, further comments will follow soon.

General remark: please split this PR in two PR's - one introducing new menu for Editor ruler and another one that introduces new trigger point behavior. It would be easier to review and easier to fix bugs later on.

iloveeclipse reviewed (changes requested):
I will try to complete tomorrow, here some feedback on what I've managed to review today

SougandhS commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/debug/core/IJavaBreakpoint.java:
```diff
@@ -47,6 +47,12 @@ public interface IJavaBreakpoint extends IBreakpoint, ITriggerPoint {
+	 *
+	 * @since 3.23
+	 */
+	public static final int RESUME_THREAD = 3;
```
RESUME_ON_SUSPEND sounds great!

SougandhS commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/breakpoints/JavaBreakpoint.java:
```diff
@@ -705,6 +705,9 @@ protected int getJDISuspendPolicy() throws CoreException {
 		if (breakpointPolicy == IJavaBreakpoint.SUSPEND_THREAD) {
 			return EventRequest.SUSPEND_EVENT_THREAD;
 		}
+		if (breakpointPolicy == IJavaBreakpoint.RESUME_THREAD) {
```
Sure 👍

SougandhS commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIThread.java:
```diff
@@ -1456,7 +1458,10 @@ private boolean handleSuspendForBreakpointInternal(JavaBreakpoint breakpoint) {
+				if (vote == IJavaBreakpointListener.DONT_SUSPEND && policy == IJavaBreakpoint.RESUME_THREAD) {
+					resumeCondition = false;
+				}
+				if (vote == IJavaBreakpointListener.DONT_SUSPEND && policy != IJavaBreakpoint.RESUME_THREAD) {
```
Done 👍

SougandhS commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIThread.java:
```diff
@@ -1481,6 +1486,13 @@ private boolean handleSuspendForBreakpointInternal(JavaBreakpoint breakpoint) {
 		// poll listeners without holding lock on thread
 		boolean suspend = true;
 		try {
+			try {
```
Thanks for this. 👍

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/debug/ui/breakpoints/JavaBreakpointConditionEditor.java:
```diff
@@ -592,6 +593,20 @@ private void disposeViewerUndoRedoActions() {
+					fWhenChange.setText(PropertyPageMessages.JavaBreakpointConditionEditor_2);
+				}
+			}
+		} catch (CoreException e) {
```
Sorry. missed to update that earlier. Updated now

SougandhS commented:
> General remark: please split this PR in two PR's - one introducing new menu for Editor ruler and another one that introduces new trigger point behavior. It would be easier to review and easier to fix bugs later on.

Removed existing changes for toggling **_Resume Triggerpoints_** and raised it as a separate enhancement proposal 👍 
https://github.com/eclipse-jdt/eclipse.jdt.debug/issues/652

SougandhS commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/debug/core/IJavaBreakpoint.java:
```diff
@@ -47,6 +47,12 @@ public interface IJavaBreakpoint extends IBreakpoint, ITriggerPoint {
+	/**
+	 * Suspend policy constant indicating a breakpoint will not suspend the target VM when hit.
+	 *
+	 * @since 3.23
```
Updated Manifest files.

SougandhS replied:
Getting build failures due to Manifest changes 
<img width="1283" alt="image" src="https://github.com/user-attachments/assets/5dea3c34-41a0-465d-9020-d759f8240fd3" />

Is this the correct way of updating ? @iloveeclipse

iloveeclipse replied:
In theory yes, but only if you would use correct bundle id :-)
`org.eclipse.jdt.debug.core` doesn't exist.

SougandhS replied:
Here 
<img width="462" alt="image" src="https://github.com/user-attachments/assets/92ffe89a-8417-45d0-b96f-4718c7defbf5" />
org.eclipse.jdt.debug;bundle-version="[3.21.0,4.0.0)", was already set in required bundles
and jdt.debug bundle already exports debug.core
<img width="599" alt="image" src="https://github.com/user-attachments/assets/5fa0ab2c-2dd2-4cd3-8f3c-35ea0b23d957" />
So do we need this change ?

SougandhS replied:
As both packages have  `org.eclipse.jdt.debug` in its requirement section I have cleared changes I made in both Manifests.

iloveeclipse replied:
> org.eclipse.jdt.debug;bundle-version="[3.21.0,4.0.0)", was already set 

The version range is wrong, as now the bundle requires **3.23.0** minimal version

SougandhS replied:
Thanks for this knowledge :) 
Updated 👍

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/breakpoints/StandardJavaBreakpointEditor.java:
```diff
@@ -44,6 +44,7 @@ public class StandardJavaBreakpointEditor extends AbstractJavaBreakpointEditor {
 	private Button fHitCountButton;
 	private Text fHitCountText;
 	private Button fSuspendThread;
+	private Button fResumeProgram;
```
I would propose to change the name to `resumeOnSuspend` or better `resumeOnHit`  for this field.

SougandhS replied:
`resumeOnHit` 👍

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/breakpoints/StandardJavaBreakpointEditor.java:
```diff
@@ -125,12 +127,15 @@ public void modifyText(ModifyEvent e) {
+
 		SWTFactory.createLabel(composite, "", 1); // spacer //$NON-NLS-1$
-		Composite radios = SWTFactory.createComposite(composite, composite.getFont(), 2, 1, GridData.FILL_HORIZONTAL, 0, 0);
+		Composite radios = SWTFactory.createComposite(composite, composite.getFont(), 3, 1, GridData.FILL_HORIZONTAL, 0, 0);
```
I believe "Resume" shouldn't be in the "Suspend" group below. Although it is understandible why you've added it here, unlike other radio buttons, it belongs solely to the "Trigger Point" checkbox, so I would propose something like this:

![image](https://github.com/user-attachments/assets/85b160e6-4303-44e2-89fe-5dd2d461bc83)

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/propertypages/PropertyPageMessages.java:
```diff
@@ -119,4 +119,8 @@ public class PropertyPageMessages extends NLS {
 	public static String VMCapabilitiesPropertyPage_4;
 	public static String VMCapabilitiesPropertyPage_6;
 	public static String VMCapabilitiesPropertyPage_9;
+	public static String BreakpointResume;
```
`BreakpointResume` -> something like `BreakpointResumeOnHit`, similar below

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIThread.java:
```diff
@@ -1478,9 +1484,21 @@ private boolean handleSuspendForBreakpointInternal(JavaBreakpoint breakpoint) {
 		}
 
+		try {
+			if (breakpoint.getSuspendPolicy() == IJavaBreakpoint.RESUME_ON_SUSPEND && resumeCondition) {
```
I would put `resumeCondition` first, just in case the first one would throw errors, if the first one is false, it will not spam the log.

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIThread.java:
```diff
@@ -1443,6 +1443,8 @@ private boolean handleSuspendForBreakpointInternal(JavaBreakpoint breakpoint) {
 			e.printStackTrace();
 		}
+
+		boolean resumeCondition = true;
```
Would be good to explain what that is, it is not obvious.
Something like `// Condition for a trigger point that is set to resume on hit`

SougandhS replied:
Comment added

iloveeclipse commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/internal/debug/core/model/JDIThread.java:
```diff
@@ -1443,6 +1443,8 @@ private boolean handleSuspendForBreakpointInternal(JavaBreakpoint breakpoint) {
 		catch (CoreException e) {
 			e.printStackTrace();
```
It is not caused your change, but while we are here, could you please properly log this exception?

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/breakpoints/StandardJavaBreakpointEditor.java:
```diff
@@ -125,12 +127,15 @@ public void modifyText(ModifyEvent e) {
 		fSuspendThread.setLayoutData(new GridData());
 		fSuspendVM = SWTFactory.createRadioButton(radios, processMnemonics(PropertyPageMessages.JavaBreakpointPage_8), 1);
 		fSuspendVM.setLayoutData(new GridData());
+		fResumeProgram = SWTFactory.createRadioButton(radios, PropertyPageMessages.BreakpointResume, 1);
```
Will check this

SougandhS replied:
Resolved this now.

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/DebugUIMessages.java:
```diff
@@ -179,6 +179,7 @@ public class DebugUIMessages extends NLS {
 	public static String JDIModelPresentation_target_suspended;
 	public static String JDIModelPresentation_thread_filtered;
 	public static String JDIModelPresentation_uncaught_62;
+	public static String JDIModelPresentation_Resume_VM;
```
We can go with JDIModelPresentation_resume_on_hit 👍

SougandhS commented on org.eclipse.jdt.debug/model/org/eclipse/jdt/debug/core/IJavaBreakpoint.java:
```diff
@@ -47,6 +47,12 @@ public interface IJavaBreakpoint extends IBreakpoint, ITriggerPoint {
+	 *
+	 * @since 3.23
+	 */
+	public static final int RESUME_ON_SUSPEND = 3;
```
RESUME_ON_HIT is a good choice 👍

SougandhS commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/breakpoints/StandardJavaBreakpointEditor.java:
```diff
@@ -189,6 +201,7 @@ protected void setBreakpoint(IJavaBreakpoint breakpoint) throws CoreException {
 		boolean hasHitCount = false;
 		String text = Util.ZERO_LENGTH_STRING;
 		boolean suspendThread = true;
+		boolean resume = false;
```
updated 👍

SougandhS commented on org.eclipse.jdt.debug.tests/tests/org/eclipse/jdt/debug/tests/breakpoints/TriggerPointBreakpointsTests.java:
```diff
@@ -69,4 +70,47 @@ public void testTriggerPointBreakpoint() throws Exception {
+		IJavaLineBreakpoint bp2 = createLineBreakpoint(21, typeName);
+		bp1.setTriggerPoint(true);
+		bp1.setConditionEnabled(true);
+		bp1.setCondition("false"); // this wont let breakpoint to resume
```
Added. works 👍

SougandhS commented:
hi @iloveeclipse , Made all the changes suggested

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/breakpoints/StandardJavaBreakpointEditor.java:
```diff
@@ -87,18 +99,60 @@ protected Button createCheckButton(Composite parent, String text) {
+					} else {
+						fResumeOnHit.setEnabled(true);
+					}
+					if (!fSuspendThread.getSelection() && !fSuspendVM.getSelection()) {
```
The blocks with conditions repeat mostly same actions in both if/else, common code is unrelated to the if/else should be moved out.

iloveeclipse commented on org.eclipse.jdt.debug.ui/ui/org/eclipse/jdt/internal/debug/ui/breakpoints/StandardJavaBreakpointEditor.java:
```diff
@@ -87,18 +99,60 @@ protected Button createCheckButton(Composite parent, String text) {
+		fResumeOnHit.addSelectionListener(new SelectionAdapter() {
+			@Override
+			public void widgetSelected(SelectionEvent e) {
+				if (fResumeOnHit.isEnabled() && javaBpConditionEditor != null) {
```
`if (javaBpConditionEditor != null)` is common for both cases and should be extracted as first condition to check

iloveeclipse reviewed (commented):
I had some comments and changes, I will push them now.

iloveeclipse commented:
@SougandhS : please provide N&N entry.

SougandhS commented:
Thank you @iloveeclipse :) 
>@SougandhS : please provide N&N entry.

Will provide N & N 👍
